File: glibc/nptl/test-rwlock-printers.py
# Common tests for the RWLockPrinter class.
#
# Copyright (C) 2016-2021 Free Software Foundation, Inc.
# This file is part of the GNU C Library.
#
# The GNU C Library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# The GNU C Library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with the GNU C Library; if not, see
# <https://www.gnu.org/licenses/>.

import sys

from test_printers_common import *

test_source = sys.argv[1]
test_bin = sys.argv[2]
printer_files = sys.argv[3:]
printer_names = ['global glibc-pthread-locks']

try:
    init_test(test_bin, printer_files, printer_names)
    go_to_main()

    var = 'rwlock'
    to_string = 'pthread_rwlock_t'

    break_at(test_source, 'Test locking (reader)')
    continue_cmd() # Go to test_locking_reader
    test_printer(var, to_string, {'Status': 'Not acquired'})
    next_cmd()
    test_printer(var, to_string, {'Status': r'Acquired \(Read\)', 'Readers': '1'})
    next_cmd()
    test_printer(var, to_string, {'Readers': '2'})
    next_cmd()
    test_printer(var, to_string, {'Readers': '3'})

    break_at(test_source, 'Test locking (writer)')
    continue_cmd() # Go to test_locking_writer
    test_printer(var, to_string, {'Status': 'Not acquired'})
    next_cmd()
    thread_id = get_current_thread_lwpid()
    test_printer(var, to_string, {'Status': r'Acquired \(Write\)',
                                  'Writer ID': thread_id})

    continue_cmd() # Exit

except (NoLineError, pexpect.TIMEOUT) as exception:
    print('Error: {0}'.format(exception))
    result = FAIL

else:
    print('Test succeeded.')
    result = PASS

exit(result)

File: glibc/nptl/test-rwlock-printers.c
/* Helper program for testing the pthread_rwlock_t pretty printer.

   Copyright (C) 2016-2021 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, see
   <https://www.gnu.org/licenses/>.  */

/* Keep the calls to the pthread_* functions on separate lines to make it easy
   to advance through the program using the gdb 'next' command.  */

#include <pthread.h>

#define PASS 0
#define FAIL 1

static int test_locking_reader (pthread_rwlock_t *rwlock);
static int test_locking_writer (pthread_rwlock_t *rwlock);

int
main (void)
{
  pthread_rwlock_t rwlock;

  int result = FAIL;

  if (test_locking_reader (&rwlock) == PASS
      && test_locking_writer (&rwlock) == PASS)
    result = PASS;
  /* Else, one of the pthread_rwlock* functions failed.  */

  return result;
}

/* Tests locking the rwlock multiple times as a reader.  */
static int
test_locking_reader (pthread_rwlock_t *rwlock)
{
  int result = FAIL;

  if (pthread_rwlock_init (rwlock, NULL) == 0
      && pthread_rwlock_rdlock (rwlock) == 0 /* Test locking (reader).  */
      && pthread_rwlock_rdlock (rwlock) == 0
      && pthread_rwlock_rdlock (rwlock) == 0
      && pthread_rwlock_unlock (rwlock) == 0
      && pthread_rwlock_unlock (rwlock) == 0
      && pthread_rwlock_unlock (rwlock) == 0
      && pthread_rwlock_destroy (rwlock) == 0)
    result = PASS;

  return result;
}

/* Tests locking the rwlock as a writer.  */
static int
test_locking_writer (pthread_rwlock_t *rwlock)
{
  int result = FAIL;

  if (pthread_rwlock_init (rwlock, NULL) == 0
      && pthread_rwlock_wrlock (rwlock) == 0 /* Test locking (writer).  */
      && pthread_rwlock_unlock (rwlock) == 0
      && pthread_rwlock_destroy (rwlock) == 0)
    result = PASS;

  return result;
}
